src/Workspaces/Core/Portable/Workspace/Solution/SourceGeneratedDocument.cs

```diff
@@ -18,6 +20,9 @@ internal SourceGeneratedDocument(Project project, SourceGeneratedDocumentState s
 
         // TODO: make this public. Tracked by https://github.com/dotnet/roslyn/issues/50546
         internal ISourceGenerator SourceGenerator => State.SourceGenerator;
+
+        internal Type SourceGeneratorType => Diagnostics.AnalyzerFileReference.UnwrapTypeIfRequired(State.SourceGenerator.GetType());
```

Comment:
So I already have another change in another PR which switches us to holding onto type names only:

https://github.com/dotnet/roslyn/pull/52094/commits/db6c8c45565f83537859707d1f272694710de2da

So anticipate merge conflicts at some point.

Reply:
I'd also generally expect the unwrap method to be called much earlier, since other stuff in the middle may also try using that type. I worry about things like this:

So I'd expect that the "unwrap" method is being called as early as possible, because consider code like this:

https://github.com/dotnet/roslyn/blob/275f4c790945c73ca308716ba56c224a95e13ca9/src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CompilationTracker.cs#L917

Although I guess if the underlying type is generic, that might still work...maybe?